File: ios/Classes/Renderer/GridRenderer.swift
import Foundation

final class GridRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.config.showGrid else { return }

    let context = frame.context
    context.setStrokeColor(frame.theme.gridColor)
    context.setLineWidth(1)

    let main = frame.mainChartRect
    let horizontalLines = 4
    for step in 0...horizontalLines {
      let y = main.minY + main.height / CGFloat(horizontalLines) * CGFloat(step)
      context.move(to: CGPoint(x: main.minX, y: y))
      context.addLine(to: CGPoint(x: main.maxX, y: y))
    }

    if let volume = frame.volumeRect {
      context.move(to: CGPoint(x: volume.minX, y: volume.minY))
      context.addLine(to: CGPoint(x: volume.maxX, y: volume.minY))
      context.move(to: CGPoint(x: volume.minX, y: volume.maxY))
      context.addLine(to: CGPoint(x: volume.maxX, y: volume.maxY))
    }

    let verticalLines = 4
    for step in 0...verticalLines {
      let x = main.minX + main.width / CGFloat(verticalLines) * CGFloat(step)
      context.move(to: CGPoint(x: x, y: main.minY))
      context.addLine(to: CGPoint(x: x, y: main.maxY))
      if let volume = frame.volumeRect {
        context.move(to: CGPoint(x: x, y: volume.minY))
        context.addLine(to: CGPoint(x: x, y: volume.maxY))
      }
    }

    context.strokePath()
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/renderer/GridRenderer.kt
package com.tradechart.plugin.renderer

import com.tradechart.plugin.engine.ChartFrame

class GridRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        if (!frame.config.showGrid) {
            return
        }

        val main = frame.mainChartRect
        val volume = frame.volumeRect
        val horizontalLines = 4
        for (step in 0..horizontalLines) {
            val y = main.top + (main.height() / horizontalLines) * step
            frame.canvas.drawLine(main.left, y, main.right, y, frame.theme.gridPaint)
        }

        if (volume != null) {
            frame.canvas.drawLine(volume.left, volume.top, volume.right, volume.top, frame.theme.gridPaint)
            frame.canvas.drawLine(volume.left, volume.bottom, volume.right, volume.bottom, frame.theme.gridPaint)
        }

        val verticalLines = 4
        for (step in 0..verticalLines) {
            val x = main.left + (main.width() / verticalLines) * step
            frame.canvas.drawLine(main.left.coerceAtMost(x), main.top, x, main.bottom, frame.theme.gridPaint)
            if (volume != null) {
                frame.canvas.drawLine(x, volume.top, x, volume.bottom, frame.theme.gridPaint)
            }
        }
    }
}
